src/btree/bt_read.c

```diff
@@ -65,261 +65,79 @@ __row_instantiate(
 }
 
 /*
- * __las_page_instantiate_verbose --
- *     Create a verbose message to display at most once per checkpoint when performing a lookaside
- *     table read.
+ * __create_birthmark_upd --
+ *     Create a birthmark update to be put on the page.
  */
-static void
-__las_page_instantiate_verbose(WT_SESSION_IMPL *session, uint64_t las_pageid)
+static int
+__create_birthmark_upd(WT_SESSION_IMPL *session, WT_BIRTHMARK_DETAILS *birthmarkp, WT_UPDATE **updp)
 {
-    WT_CACHE *cache;
-    uint64_t ckpt_gen_current, ckpt_gen_last;
+    WT_ITEM empty_item;
+    WT_UPDATE *upd;
+    size_t incr;
 
-    if (!WT_VERBOSE_ISSET(session, WT_VERB_LOOKASIDE | WT_VERB_LOOKASIDE_ACTIVITY))
-        return;
+    *updp = NULL;
+    empty_item.size = 0;
 
-    cache = S2C(session)->cache;
-    ckpt_gen_current = __wt_gen(session, WT_GEN_CHECKPOINT);
-    ckpt_gen_last = cache->las_verb_gen_read;
+    WT_RET(__wt_update_alloc(session, &empty_item, &upd, &incr, WT_UPDATE_BIRTHMARK));
+    upd->txnid = birthmarkp->txnid;
+    upd->durable_ts = birthmarkp->durable_ts;
+    upd->start_ts = birthmarkp->start_ts;
+    upd->prepare_state = birthmarkp->prepare_state;
+    *updp = upd;
 
-    /*
-     * This message is throttled to one per checkpoint. To do this we track the generation of the
-     * last checkpoint for which the message was printed and check against the current checkpoint
-     * generation.
-     */
-    if (WT_VERBOSE_ISSET(session, WT_VERB_LOOKASIDE) || ckpt_gen_current > ckpt_gen_last) {
-        /*
-         * Attempt to atomically replace the last checkpoint generation for which this message was
-         * printed. If the atomic swap fails we have raced and the winning thread will print the
-         * message.
-         */
-        if (__wt_atomic_casv64(&cache->las_verb_gen_read, ckpt_gen_last, ckpt_gen_current)) {
-            __wt_verbose(session, WT_VERB_LOOKASIDE | WT_VERB_LOOKASIDE_ACTIVITY,
-              "Read from lookaside file triggered for "
-              "file ID %" PRIu32 ", page ID %" PRIu64,
-              S2BT(session)->id, las_pageid);
-        }
-    }
+    return (0);
 }
 
 /*
- * __las_page_instantiate --
- *     Instantiate lookaside update records in a recently read page.
+ * __instantiate_birthmarks --
+ *     Instantiate birthmarks records in a recently read page.
  */
 static int
-__las_page_instantiate(WT_SESSION_IMPL *session, WT_REF *ref)
+__instantiate_birthmarks(WT_SESSION_IMPL *session, WT_REF *ref)
```

Comment:
What's happening here? Just checking my mental model is correct.

We have some data about the birthmark for each key attached to the `WT_REF` regardless of whether the page is in cache or not. When we read the page into cache, we're calling this function and populating an update list (?) with the birthmark records using that data.

Is that right?

Reply:
You are correct except:
1. We do not have a birthmark per key. Some keys might not have any birthmark.
2. Only pages that were previously lookaside evicted from the cache will have these birthmarks. These birthmarks are used to create on-page updates when this lookaside evicted page is read back into memory. We free these after a successful read. So, a page in cache should not continue having these birthmark/key pairs attached to ref.